hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ObservableResultsTest.java

```diff
@@ -0,0 +1,240 @@
+
+    @Before
+    public void before() {
+        observableName = randomName();
+        testObserver = new TestObserver();
+        testObservable = jet().getObservable(observableName);
+        testObservable.addObserver(testObserver);
+    }
+
+    @After
+    public void after() {
+        testObservable.destroy();
+    }
+
+    @Test
+    public void batchJobCompletesSuccessfully() {
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+        job.join();
+
+        assertSortedValues(testObserver, 0L, 1L, 2L, 3L, 4L);
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 1);
+    }
+
+    @Test
+    public void batchJobFails() {
+        BatchSource<String> errorSource = SourceBuilder
+                .batch("error-source", x -> (Object) null)
+                .<String>fillBufferFn((in, Void) -> {
+                    throw new Exception("Ooops!");
+                })
+                .destroyFn(ConsumerEx.noop())
+                .build();
+
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(errorSource)
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+        assertTrueEventually(() -> assertEquals(JobStatus.FAILED, job.getStatus()));
+
+        assertSortedValues(testObserver);
+        assertError(testObserver, "Ooops!");
+        assertCompletions(testObserver, 0);
+    }
+
+    @Test
+    public void streamJob() {
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(TestSources.itemStream(100))
+                .withoutTimestamps()
+                .map(SimpleEvent::sequence)
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+
+        assertTrueEventually(() -> assertTrue(testObserver.getSortedValues().size() > 10));
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 0);
+
+        job.cancel();
+        assertTrueEventually(() -> assertTrue(testObserver.getSortedValues().size() > 10));
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 0);
+    }
+
+    @Test
+    public void streamJobRestart() {
+        Pipeline pipeline = Pipeline.create();
+        pipeline.readFrom(TestSources.itemStream(100))
+                .withoutTimestamps()
+                .map(SimpleEvent::sequence)
+                .writeTo(Sinks.observable(observableName));
+
+        Job job = jet().newJob(pipeline);
+
+        assertTrueEventually(() -> assertTrue(testObserver.getSortedValues().size() > 10));
+        assertError(testObserver, null);
+        assertCompletions(testObserver, 0);
+
+        job.restart();
+        assertTrueEventually(() -> assertEquals(JobStatus.RUNNING, job.getStatus()));
```

Comment:
It can be valuable here to check whether observer gets next values also after job restart.